nuxeo-drive-client/nxdrive/tests/test_engine_offline.py

```diff
@@ -0,0 +1,132 @@
+import nxdrive
+from mock import patch
+from urllib2 import URLError
+from time import sleep
+from nxdrive.client.base_automation_client import BaseAutomationClient
+
+server_online = True
+original_execute = BaseAutomationClient.execute
+original_fetch_api = BaseAutomationClient.fetch_api
+
+
+class EngineOffLineTestCase(UnitTestCase):
+
+    '''
+    1. fetch_api() and execute() methods in BaseAutomationClient are mocked to simulate network error
+    2. Files/folders should not sync while engine is paused due to a network error
+    3. Files/folders should not sync on manual pause and should sync on resume.
+    '''
+
+    def setUp(self):
+        super(EngineOffLineTestCase, self).setUp()
+        # Start engine and wait for Sync
+        self.engine_1.start()
+        self.wait_sync(wait_for_async=30)
+
+    def tearDown(self):
+        pass
+
+    def mock_fetch_api(self):
+        global server_online
+        if not server_online:
+            raise URLError("server is offline in mock_fetch_api:")
+        return original_fetch_api(self)
+
+    def mock_execute(self, *args, **kwargs):
+        global server_online
+        if not server_online and ('NuxeoDrive' in args[0] or 'Workspace' in args[0] or args[0].strip('/').endswith('/automation')):
+            # block all Nuxeo Drive APIs and also UserWorkspace.Get API only when server offline
+            raise URLError("server is offline in mock_execute:")
+        return original_execute(self, *args, **kwargs)
+
+    @patch.object(nxdrive.client.base_automation_client.BaseAutomationClient, 'fetch_api', mock_fetch_api)
+    @patch.object(nxdrive.client.base_automation_client.BaseAutomationClient, 'execute', mock_execute)
+    def test_with_mock_server_offline(self):
+
+        '''
+        1. Simulate Network Error
+        2. Create a local folder inside user folder
+        3. Create a file inside the folder
+        4. Check the pair_state and error_count for the folder/files when engine goes offline
+        5. Call original method to resume engine
+        4. Check the pair_state when engine goes online
+        '''
+
+        global server_online
+        server_online = False
+        # Wait for GetChangeSummary to call check_offline method
+        sleep(35)
+
+        # Create a folder inside user folder and a file inside the folder
+        self.local_client_1.make_folder(u'/', 'NetworkError')
+        self.local_client_1.make_file(u'/NetworkError', 'NetworkErrorFile.txt', content="test network failure")
+        # Wait for events to be handled
+        sleep(35)
+
+        # Check pair_state and error_count for the folder/file because it should not try to sync when engine is offline
+        test_folder = self.get_dao_state_from_engine_1('/NetworkError')
+        self.assertEqual(test_folder.pair_state, 'locally_created')
+        self.assertEqual(test_folder.error_count, 0)
+        test_file = self.get_dao_state_from_engine_1('/NetworkError/NetworkErrorFile.txt')
```

Comment:
this helper test method here is problematic. I tested with cpo-dev1 stack and users 'dsuser1', 'dsuser2' (both with pwd 'pycon2016') and this method uses the workspace_title_1 which is the user's full name... and it fails.
I used instead:
test_file = self.engine_1.get_dao().get_state_from_local('/My Docs/NetworkErrorFile.txt')
and it works.